### DISCUSSION TITLE
Power BI custom connector application logs not appearing

### DISCUSSION THREAD

**Discussion by arindamroy19** [microsoft/vscode-powerquery-sdk#297] [Q&A]
I am building a PowerBI custom connector for Zoho with OAuth2 flow using the Power Query SDK for VSCode. I want to emit debug/trace statements from the connector, and am using the following -

`logMessage3 = Diagnostics.Trace(TraceLevel.Warning, "TextValueFromNumber", "In Contents", true),
`

But, these logs don't appear in log files in the 'Power BI Desktop Store App\Traces' location. I only see PowerBI generic log statements. In Visual Studio Projects, there is a flag 'Show User Traces' - https://learn.microsoft.com/en-us/power-query/samples/trippin/8-diagnostics/readme

Is there something similar for VSCode SDK Projects?

**Comment by bgribaudo** [comment]
Hi @arindamroy19!

To get the SDK to output logs, set the `-l` / `--logMashupEngineTraces` option when calling `PQTest run-test`.  

`-l` can be set to:
* `User` for just user traces to be output (e.g. traces you create using `Diagnostics.Trace`)
* `Engine` for just engine traces to be output
* `All` for both of the about categories of traces to be output

The specified traces will be written to a logfile whose location is given in the output of `PQTest run-test` in property `TraceFileLocation`.

For example, you could run something like the following, then look in the output produced for the `TraceFileLocation`:
````
PQTest.exe run-test --extension "SomeExtension.mez" --prettyPrint --queryFile .\SomeExtension.query.pq --logMashupEngineTraces User
````

**Reply by arindamroy19** [reply]
Thanks @bgribaudo. This seems to be for the PQTest utility. What about from within the context of PowerBI after the connector has been deployed?

**Reply by mattmasson** [reply]
See:

- https://learn.microsoft.com/en-us/power-bi/fundamentals/desktop-diagnostics#collecting-mashup-traces
- https://learn.microsoft.com/en-us/power-query/query-diagnostics
- https://learn.microsoft.com/en-us/data-integration/gateway/service-gateway-tshoot#collect-logs-from-the-on-premises-data-gateway-app

**Comment by bgribaudo** [comment]
@mattmasson, is there a way to set some configuration value so that the IDE sets the log traces option when it runs tests?

**Reply by mattmasson** [reply]
You can create a custom task and pass command line parameters using the `additionalArgs` field. 

If you go to **Terminal | Run Task ...**, you should see a `powerquery` folder with our build-in tasks. From there you can select the gear icon for one of the tasks to customize it. 

It would look something like:

```json
{
  "type": "powerquery",
  "operation": "run-test",
  "pathToConnector": "${config:powerquery.sdk.defaultExtension}",
  "pathToQueryFile": "${workspaceFolder}/mytestfile.query.pq",
  "additionalArgs": ["--logMashupEngineTraces", "User"],
  "problemMatcher": [],
  "label": "Run test with tracing"
}
```

It would be nice to have an extension level setting to always run with tracing on, but I think we'd also need some sort of cleanup functionality to go along with it.

PQTest will output the trace files to your `%localappdata%\Microsoft\pqtest\Logs` directory.

**Reply by bgribaudo** [reply]
Thanks, Matt!

**Comment by mattmasson** [comment]
@arindamroy19, also note that M statements are lazily evaluated, and query steps that aren't referenced/needed never actually get evaluated. So in your example, the diagnostics message would never be output unless another part of your query is actually trying to get the value of `logMessage3`. 

https://learn.microsoft.com/en-us/power-query/samples/trippin/8-diagnostics/readme
